sklearn/ensemble/bagging.py

```diff
@@ -276,27 +278,47 @@ def fit(self, X, y, sample_weight=None):
             raise ValueError("Out of bag estimation only available"
                              " if bootstrap=True")
 
-        # Free allocated memory, if any
-        self.estimators_ = None
+        if not self.warm_start:
+            # Free allocated memory, if any
+            self.estimators_ = []
+            self.estimators_samples_ = []
+            self.estiamtors_features_ = []
 
-        # Parallel loop
-        n_jobs, n_estimators, starts = _partition_estimators(self.n_estimators,
-                                                             self.n_jobs)
-        seeds = random_state.randint(MAX_INT, size=self.n_estimators)
+        n_more_estimators = self.n_estimators - len(self.estimators_)
 
-        all_results = Parallel(n_jobs=n_jobs, verbose=self.verbose)(
-            delayed(_parallel_build_estimators)(
-                n_estimators[i],
-                self,
-                X,
-                y,
-                sample_weight,
-                seeds[starts[i]:starts[i + 1]],
-                verbose=self.verbose)
-            for i in range(n_jobs))
+        if n_more_estimators < 0:
+            raise ValueError('n_estimators=%d must be larger or equal to '
+                             'len(estimators_)=%d when warm_start==True'
+                             % (self.n_estimators, len(self.estimators_)))
+
+        elif n_more_estimators == 0:
+            warn("Warm-start fitting without increasing n_estimators does not "
+                 "fit new trees.")
+        else:
+            # Parallel loop
+            n_jobs, n_estimators, starts = _partition_estimators(n_more_estimators,
+                                                                 self.n_jobs)
+
+            # Advance random state to state after training
+            # the first n_estimators
+            if self.warm_start and len(self.estimators_) > 0:
+                random_state.randint(MAX_INT, size=len(self.estimators_))
+
+            seeds = random_state.randint(MAX_INT, size=n_more_estimators)
+
+            all_results = Parallel(n_jobs=n_jobs, verbose=self.verbose)(
+                delayed(_parallel_build_estimators)(
+                    n_estimators[i],
+                    self,
+                    X,
+                    y,
+                    sample_weight,
+                    seeds[starts[i]:starts[i + 1]],
+                    verbose=self.verbose)
+                for i in range(n_jobs))
 
         # Reduce
-        self.estimators_ = list(itertools.chain.from_iterable(
+        self.estimators_ += list(itertools.chain.from_iterable(
             t[0] for t in all_results))
         self.estimators_samples_ = list(itertools.chain.from_iterable(
```

Comment:
Why not updating  `self.estimators_samples_`here instead of earasing the whole  attribute.